docs: Formatters page updates

bpmutter opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [JS Foundation Code of Conduct](https://eslint.org/conduct).
-->

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

<!--
    The following template is intentionally not a markdown checkbox list for the reasons
    explained in https://github.com/eslint/eslint/pull/12848#issuecomment-580302888
-->

[x] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)

Miscellaneous changes on [Formatters](https://eslint.org/docs/latest/user-guide/formatters/) page:

- In the example source section, make fullOfProblems.js and .eslintrc: not headers.
- For the Example Outputs section, provide an example of the CLI input, and also give a few sentence description of what each of the output is, and why you might want to use it.
- Other small cosmetic changes and refactors

#### Is there anything you'd like reviewers to focus on?

<!-- markdownlint-disable-file MD004 -->

Fixes #16476

mdjermanovic commented on docs/src/user-guide/formatters/index.md:
````diff
@@ -83,28 +97,48 @@ function addOne(i) {
 {&#34;results&#34;:[{&#34;filePath&#34;:&#34;/var/lib/jenkins/workspace/Releases/eslint Release/eslint/fullOfProblems.js&#34;,&#34;messages&#34;:[{&#34;ruleId&#34;:&#34;no-unused-vars&#34;,&#34;severity&#34;:2,&#34;message&#34;:&#34;&#39;addOne&#39; is defined but never used.&#34;,&#34;line&#34;:1,&#34;column&#34;:10,&#34;nodeType&#34;:&#34;Identifier&#34;,&#34;messageId&#34;:&#34;unusedVar&#34;, ...
 ```
 
 ### json
 
+TODO: quick overview
````
This one just outputs JSON-serialized lint results. Useful when you want to do something with it programmatically, because it's structured data.

Context: https://github.com/eslint/eslint/issues/3036

mdjermanovic commented on docs/src/user-guide/formatters/index.md:
````diff
@@ -83,28 +97,48 @@ function addOne(i) {
 &lt;?xml version=&#34;1.0&#34; encoding=&#34;utf-8&#34;?&gt;&lt;jslint&gt;&lt;file name=&#34;/var/lib/jenkins/workspace/Releases/eslint Release/eslint/fullOfProblems.js&#34;&gt;&lt;issue line=&#34;1&#34; char=&#34;10&#34; evidence=&#34;&#34; reason=&#34;&amp;apos;addOne&amp;apos; is defined but never used. (no-unused-vars)&#34; /&gt;&lt;issue line=&#34;2&#34; char=&#34;9&#34; evidence=&#34;&#34;  ...
 ```
 
 ### json-with-metadata
 
+TODO: quick overview
````
Like JSON, but with additional data.

Context: https://github.com/eslint/rfcs/pull/10

nzakas replied:
We did this to avoid breaking changes to the `json` formatter. Adding a bunch of new data in there would have broken existing consumers.

mdjermanovic commented on docs/src/user-guide/formatters/index.md:
```diff
@@ -145,6 +183,10 @@ function addOne(i) {
 
 ### tap
 
+TODO: quick overview
```
https://testanything.org/

When you want to pass the results to a system that expects this format.

Context: , https://github.com/eslint/eslint/issues/608

mdjermanovic commented on docs/src/user-guide/formatters/index.md:
```diff
@@ -211,6 +253,10 @@ not ok 1 - /var/lib/jenkins/workspace/Releases/eslint Release/eslint/fullOfProbl
 
 ### unix
 
+TODO: quick overview
```
This is another common format.

Context: https://github.com/eslint/eslint/issues/2991

mdjermanovic commented on docs/src/user-guide/formatters/index.md:
````diff
@@ -83,28 +97,48 @@ function addOne(i) {
 [{&#34;filePath&#34;:&#34;/var/lib/jenkins/workspace/Releases/eslint Release/eslint/fullOfProblems.js&#34;,&#34;messages&#34;:[{&#34;ruleId&#34;:&#34;no-unused-vars&#34;,&#34;severity&#34;:2,&#34;message&#34;:&#34;&#39;addOne&#39; is defined but never used.&#34;,&#34;line&#34;:1,&#34;column&#34;:10,&#34;nodeType&#34;:&#34;Identifier&#34;,&#34;messageId&#34;:&#34;unusedVar&#34;,&#34;endLine&#34;:1 ...
 ```
 
 ### junit
 
+TODO: quick overview
````
Another one that you use when you want to pass the data to a system that expects a specific format.

Context: https://github.com/eslint/eslint/issues/16

nzakas replied:
Meant for use with the [JUnit Jenkins plugin](https://plugins.jenkins.io/junit/).

mdjermanovic commented on docs/src/user-guide/formatters/index.md:
```diff
@@ -125,6 +159,10 @@ function addOne(i) {
 
 ### stylish
 
+TODO: quick overview
```
Readable output for CLI users. This is the default formatter.

mdjermanovic commented on docs/src/user-guide/formatters/index.md:
```diff
@@ -57,16 +57,30 @@ function addOne(i) {
+
+## Built-In Formatter Options
 
 ### checkstyle
 
+TODO: quick overview
```
Another one to pass the data to a system that expects a specific format,

Context: https://github.com/eslint/eslint/issues/14 (not much context there though, here's [checkstyle docs](https://checkstyle.sourceforge.io/)).

nzakas replied:
Back in the day, a lot of Jenkins plugins expected results in Checkstyle format. It's not nearly as popular as it was but there are still a few Jenkins plugins that expect that format:

https://plugins.jenkins.io/ui/search?query=checkstyle

mdjermanovic commented on docs/src/user-guide/formatters/index.md:
```diff
@@ -83,28 +97,48 @@ function addOne(i) {
+
 <iframe src="html-formatter-example.html" width="100%" height="460px"></iframe>
 
 ### jslint-xml
 
+TODO: quick overview
```
There's not much context in https://github.com/eslint/eslint/issues/15, but I'm pretty sure this is another one you'd use if you need to pass data to a system that expects a particular format.

nzakas replied:
There is a [JSLint plugin for Jenkins](https://plugins.jenkins.io/jslint/) that expects a specific XML format to produce reports. This formatter type outputs data in the format that the Jenkins plugin expects. This was necessary so that people could swap out JSLint for ESLint in their Jenkins projects.

mdjermanovic commented on docs/src/user-guide/formatters/index.md:
```diff
@@ -227,6 +273,10 @@ not ok 1 - /var/lib/jenkins/workspace/Releases/eslint Release/eslint/fullOfProbl
 
 ### visualstudio
 
+TODO: quick overview
```
Explained in https://github.com/eslint/eslint/issues/4708#issue-122267920

mdjermanovic commented on docs/src/user-guide/formatters/index.md:
```diff
@@ -83,28 +97,48 @@ function addOne(i) {
 
 ### html
 
+TODO: quick overview
```
For a nice visual presentation in a browser.

mdjermanovic commented on docs/src/user-guide/formatters/index.md:
````diff
@@ -57,16 +57,30 @@ function addOne(i) {
 &lt;?xml version=&#34;1.0&#34; encoding=&#34;utf-8&#34;?&gt;&lt;checkstyle version=&#34;4.3&#34;&gt;&lt;file name=&#34;/var/lib/jenkins/workspace/Releases/eslint Release/eslint/fullOfProblems.js&#34;&gt;&lt;error line=&#34;1&#34; column=&#34;10&#34; severity=&#34;error&#34; message=&#34;&amp;apos;addOne&amp;apos; is defined but never used. (no-unused-vars)&#34; source=&#34;eslint.rules.no-unused- ...
 ```
 
 ### compact
 
+TODO: quick overview
````
This one was added in the [initial commit](https://github.com/eslint/eslint/commit/a658d7b0e7d915750f18d666823d54ef2129a9af#diff-b3a58c91578d9380ce8b33e6f165445c2ec7352033f5cfca85d36930d71b8daf), so there's no context to find. I believe it's primarily intended to be readable output for CLI users, but will ask @nzakas to confirm.

nzakas replied:
This format mimics the default output from JSHint, and yes, is meant to be easier for humans to read.

nzakas commented on docs/src/user-guide/formatters/index.md:
````diff
@@ -42,7 +42,7 @@ function addOne(i) {
 ```
 
-### `.eslintrc`
+`.eslintrc`:
````
While we're at it, let's use a file extension:

```suggestion
`.eslintrc.json`:
```

nzakas reviewed (commented):
Note: This page is autogenerated. Any changes made here will go live but will be overwrriten the next time the site is generated. The evergreen changes should be made in this file:

https://github.com/eslint/eslint/blob/main/templates/formatter-examples.md.ejs

bpmutter commented:
@nzakas and @mdjermanovic thank you for the context! i'll try to fill out all the descriptions with this information. will let you know in comments if need any further clarification here.

and regarding:

> Note: This page is autogenerated. Any changes made here will go live but will be overwrriten the next time the site is generated. The evergreen changes should be made in this file:
>
> https://github.com/eslint/eslint/blob/main/templates/formatter-examples.md.ejs

at first glance, i understand what's going on with the template conceptually, but am not sure how i'd add the descriptions to it. will dig in a bit, and let you know if any questions. 

if i cannot figure out how to add the descriptions to the template, i'll add them to the markdown file, so we can at least get the copy down before later adding it to the template file.

bpmutter commented:
i've updated the code to include descriptions in the templates by adding a `_manifest.js` file to the source code, which contains the descriptions.

i've also:
- updated the template to include descriptions and changes that were previously made to `docs/src/user-guide/formatters/index.md`. 
- reverted the changes to `docs/src/user-guide/formatters/index.md` since the page is generated.

generate the output as expected, you can run `npm run gensite`.

i'm not convinced that using the `_manifest.js` file as is done here is the best approach for including descriptions from the source code, even if it is effective. i look forward to the maintainer team's feedback on if/how to refactor this.

nzakas commented on lib/cli-engine/formatters/_manifest.js:
```diff
@@ -0,0 +1,64 @@
+/**
```
Because this is just data, can we just create a JSON file here? I'd call it `formatter-data.json` and you'd still be able to `require()` it, but would otherwise be ignored.

bpmutter replied:
sgtm, i like this approach

bpmutter commented:
@nzakas implemented w the file named `formatter-meta.json` to align w the file naming convention of `rules-meta.json`

nzakas reviewed (approved):
LGTM. Would like @mdjermanovic to review before merging.

mdjermanovic commented on lib/cli-engine/formatters/formatters-meta.json:
```diff
@@ -0,0 +1,46 @@
+    },
+    {
+        "name": "json-with-metadata",
+        "description": "Outputs JSON-serialized results. The `json-with-meta` provides the same linting results as the [`json`](#json) formatter with additional metadata about the rules applied. The linting results are included in the `results` property and the rules metadata is included in the `metadata` property.\n\nAlternatively, you can use the [ESLint Node.js API](../../developer-guide/nodej ...
```
```suggestion
        "description": "Outputs JSON-serialized results. The `json-with-metadata` provides the same linting results as the [`json`](#json) formatter with additional metadata about the rules applied. The linting results are included in the `results` property and the rules metadata is included in the `metadata` property.\n\nAlternatively, you can use the [ESLint Node.js API](../../developer-guide/nodejs-api) to programmatically use ESLint."
```

mdjermanovic commented on lib/cli-engine/formatters/formatters-meta.json:
```diff
@@ -0,0 +1,46 @@
+    },
+    {
+        "name": "json",
+        "description": "Outputs JSON-serialized results. The `json` formatter is useful when you want to programmatically work with the CLI's linting results.\n\nAlternatively, you can use the [ESLint Node.js API](../../developer-guide/nodejs-api.md) to programmatically use ESLint."
```
```suggestion
        "description": "Outputs JSON-serialized results. The `json` formatter is useful when you want to programmatically work with the CLI's linting results.\n\nAlternatively, you can use the [ESLint Node.js API](../../developer-guide/nodejs-api) to programmatically use ESLint."
```

mdjermanovic commented on templates/formatter-examples.md.ejs:
```diff
@@ -22,7 +22,7 @@ The built-in formatter options are:
 
 Examples of each formatter were created from linting `fullOfProblems.js` using the `.eslintrc` configuration shown below.
```
```suggestion
Examples of each formatter were created from linting `fullOfProblems.js` using the `.eslintrc.json` configuration shown below.
```

mdjermanovic commented on templates/formatter-examples.md.ejs:
````diff
@@ -49,17 +49,26 @@ function addOne(i) {
+Tests the formatters with the CLI:
+
+```shell
+npx eslint --formatter <Add formatter here> fullOfProblems.js
````
```suggestion
npx eslint --format <Add formatter here> fullOfProblems.js
```

bpmutter commented:
@mdjermanovic ready for re-review!
